feat(lint): added lib for getting corresponding yaml node for a given schema struct from a skaffold config

aaron-prindle opened the pull request:
feat: added lib for getting corresponding yaml node for a given schema struct from a skaffold config

briandealwis reviewed (commented):
I haven't look in detail at the implementation but this approach looks promising.

I think you want to try to think of the API to get the file+line+column+... without having to return kyaml objects.

It would make sense to integrate this directly into `pkg/skaffold/parser`, wouldn't it?  And it would be very useful to track updates from applying profiles (e.g., map artifact A that was defined in a profile to the location of the A in the profile definition).  I'm not sure what we do with patch profiles that change fields within an object — [such as changing an artifact's name](https://github.com/GoogleContainerTools/container-debug-support/blob/duct-tape/nodejs/skaffold.yaml#L71-L80).  Maybe we need a way to record enough information to report "artifact <x> (as altered by profile <Y>)".

aaron-prindle commented on pkg/skaffold/schema/validation/validation.go:
```diff
@@ -493,10 +520,17 @@ func validatePortForwardResources(pfrs []*latestV1.PortForwardResource) []error
 		resourceType := strings.ToLower(string(pfr.Type))
 		if _, ok := validResourceTypes[resourceType]; !ok {
-			errs = append(errs, fmt.Errorf("%s is not a valid resource type for port forwarding", pfr.Type))
+			portForwardYAMLInfo, err := yamlnodes.Manager.Get(pfrs[i])
```
This makes a lot of sense I will update the logic to have a `Locate` and investigate the `iterator`.  

One thing to consider with only returning a location is that currently this method only supports line numbers for structs, arrays and slices which we can reflect to get a pointer to.  For example to get the line number for something like:
`config.Build.Artifacts[0].Name`  (type string) is not possible to do directly with this method, only `config.Build.Artifacts[0]` will be in the map. Getting a primitive field for a struct requires additional processing of the yaml Node and this is why I was using nodes vs line numbers. That way a user could get the node for `config.Build.Artifacts[0]` and then query for the value they want eg: `artifactNode.Pipe(kyaml.Lookup("name")`.  I agree definitely makes sense to in the worst case at least support both methods but I am curious if you have ideas on how to support the primitive types w/ an approach like this?  We could have some kind of additional `LocateChild` or `LocateSubField` method that takes a struct and the name of the field and return a line #, eg:
```
// LocateSubField returns the file and location of the provided structure and field of the structure, a pointer to an object defined by this config as well as the name of the SubField.
// filename is empty and line, col are -1 if not found.
func (*MyConfig) LocateSubField(obj interface{}, subField string) (filename string, line, col, length int)
```

aaron-prindle replied:
Added `Locate` API, still working through how struct fields should be queried

aaron-prindle commented:
Updated the API a bit here to be a part of `pkg/skaffold/parser` and expose `Location` objects vs *kyaml.Node objects.  Also added line ending information to `Location` obj (StartLine, EndLine, StartColumn, EndColumn)

aaron-prindle commented:
> I'm not sure what we do with patch profiles that change fields within an object — [such as changing an artifact's name](https://github.com/GoogleContainerTools/container-debug-support/blob/duct-tape/nodejs/skaffold.yaml#L71-L80). Maybe we need a way to record enough information to report "artifact (as altered by profile )".

Yes, I'm still figuring out what to do with patch profiles (and profiles generally) as well here.  I think that we will need to track how config objects are modified via patches & profile patches as you state

aaron-prindle commented on pkg/skaffold/yamlnodes/yamlnodes.go:
```diff
@@ -0,0 +1,155 @@
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+// TODO(aaron-prindle) change to better singleton model if necesssary (currently obj needs to updated on file changes so leaving out for now)
```
Done
